### PR TITLE
[TS-429] Implemented Socket Server for real time chat 

### PR DESCRIPTION
**Pull request by tungleu**
(posted a video on discord) 
To test this PR:
1. Clone the tradeshare-socket and run it 
2. Have 2 browsers for 2 accounts (test and test account) 

Scope of this PR: 
- Implement socket-io to enable real-time chat between `test` and `test account` 

What is not included (will be updated): 
- Create conversation 
- Persistent message 
- User's Name (so far we only have usernames)

### CODE REVIEW

**Review by AlyaNaseer — CHANGES_REQUESTED** [review]
So cool! Awesome work!

Like you mentioned in your video, there are some things that are not included in your PR so you might change the scope. I'd specify what is actually included and what isn't included in your PR for easy referral.

I'll list below some of the things that remain to do:
- Replace username by name in the chat window and chat menu (**)
- Have the username below the name in the chat window (*)
- Save the messages in database
- The avatar in the chat window seems to be a bit low, but it might be because the row containing the username isn't filled (*)
- Rewrite Cypress tests (**)
- Time since last message

The ones with ** are ones I recommend you do in your PR now since I think it should be required to call this PR complete (especially since you removed the Cypress tests, to make sure we don't forget to add it back, it could be good to do it now). The ones with * are ones that could be an easy fix in this PR as well but I don't consider them as absolutely necessary for the scope of this PR. And of course the DB implementation will be its own thing.

**Comment by GadoliniumPotassium** [comment]
I can't believe you made snapchat!

File: src/components/Messages/ChatMenu.vue [hunk 1]
```diff
@@ -102,162 +65,50 @@
...
+  data: function () {
+    return {
+      conversations: []
+    }
+  },
   computed: {
+    user () {
+      return JSON.parse(localStorage.getItem('user'))
+    },
+    //
```
Comment (itsShaVawn) [inline src/components/Messages/ChatMenu.vue:85]: remove this

File: src/components/Messages/Conversation.vue [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+      <v-col class="px-0 py-0">
+        <v-list-item-subtitle
+          class="pt-2 text-caption text-truncate"
+          style="max-width: 19vw;"
+          v-text="chat.latestMessage"
+        />
+      </v-col>
+    </v-list-item-content>
+  </v-list-item>
+</template>
```
Comment (itsShaVawn) [inline src/components/Messages/Conversation.vue:34]: add new line

File: src/components/Messages/Conversation.vue [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+      default: null
+    }
+  },
+  computed: {
+    user () {
+      return JSON.parse(localStorage.getItem('user'))
+    }
+  }
+}
+
```
Comment (itsShaVawn) [inline src/components/Messages/Conversation.vue:50]: remove extra line

File: src/components/Messages/Conversation.vue [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+</script>
+
+<style scoped>
```
Comment (itsShaVawn) [inline src/components/Messages/Conversation.vue:53]: remove

**Review by itsShaVawn — CHANGES_REQUESTED** [review]
In addition to what @AlyaNaseer said: 
image

I would have also liked to see the "New message" button being used in your video.

**Comment by AlyaNaseer** [comment]
> In addition to what @AlyaNaseer said: image
> 
> I would have also liked to see the "New message" button being used in your video.

I think some of the changes you ask might be difficult in this PR since the DB isn't implemented, but I'm not 100% sure

**Comment by itsShaVawn** [comment]
> > In addition to what @AlyaNaseer said: image
> > I would have also liked to see the "New message" button being used in your video.
> 
> I think some of the changes you ask might be difficult in this PR since the DB isn't implemented, but I'm not 100% sure

These have nothing to do with the DB though.

Maybe this will be more clear:

- Circle 1: profile picture and username are misaligned. If anything, the pfp should be centered and not touching the divider.
- Circle 2: the username is breaking because the container is small, but there is ample amount of space for the username to avoid wrapping.
- Circle 3: this is the only one involving the DB, where the message in the circle doesn't appear in the chat window.
- Circle 4: the height of the messaging component doesn't extend to the whole available page.

**Comment by itsShaVawn** [comment]
u have 3 code smell

**Comment by tungleu** [comment]
> u have 3 code smell

oops

**Comment by tungleu** [comment]
> So cool! Awesome work!
> 
> Like you mentioned in your video, there are some things that are not included in your PR so you might change the scope. I'd specify what is actually included and what isn't included in your PR for easy referral.
> 
> I'll list below some of the things that remain to do:
> 
> * Replace username by name in the chat window and chat menu (**)
> * Have the username below the name in the chat window (*)
> * Save the messages in database
> * The avatar in the chat window seems to be a bit low, but it might be because the row containing the username isn't filled (*)
> * Rewrite Cypress tests (**)
> * Time since last message
> 
> The ones with ** are ones I recommend you do in your PR now since I think it should be required to call this PR complete (especially since you removed the Cypress tests, to make sure we don't forget to add it back, it could be good to do it now). The ones with * are ones that could be an easy fix in this PR as well but I don't consider them as absolutely necessary for the scope of this PR. And of course the DB implementation will be its own thing.

since our current conversation model in the back end only has a username field so I'm not able to fetch the name at the moment. For the cypress test i already edited it, I just removed some static data tests

**Comment by tungleu** [comment]
> > > In addition to what @AlyaNaseer said: image
> > > I would have also liked to see the "New message" button being used in your video.
> > 
> > 
> > I think some of the changes you ask might be difficult in this PR since the DB isn't implemented, but I'm not 100% sure
> 
> These have nothing to do with the DB though.
> 
> Maybe this will be more clear:
> 
> * Circle 1: profile picture and username are misaligned. If anything, the pfp should be centered and not touching the divider.
> * Circle 2: the username is breaking because the container is small, but there is ample amount of space for the username to avoid wrapping.
> * Circle 3: this is the only one involving the DB, where the message in the circle doesn't appear in the chat window.
> * Circle 4: the height of the messaging component doesn't extend to the whole available page.



> > > In addition to what @AlyaNaseer said: image
> > > I would have also liked to see the "New message" button being used in your video.
> > 
> > 
> > I think some of the changes you ask might be difficult in this PR since the DB isn't implemented, but I'm not 100% sure
> 
> These have nothing to do with the DB though.
> 
> Maybe this will be more clear:
> 
> * Circle 1: profile picture and username are misaligned. If anything, the pfp should be centered and not touching the divider.
> * Circle 2: the username is breaking because the container is small, but there is ample amount of space for the username to avoid wrapping.
> * Circle 3: this is the only one involving the DB, where the message in the circle doesn't appear in the chat window.
> * Circle 4: the height of the messaging component doesn't extend to the whole available page.

image
temporary removed latest message as this was actually fetch from the backend and i didnt write the new messages into the database and also replace user's name with user's username so that the profile pic wont overflow

**Comment by AlyaNaseer** [comment]
> > > In addition to what @AlyaNaseer said: image
> > > I would have also liked to see the "New message" button being used in your video.
> > 
> > 
> > I think some of the changes you ask might be difficult in this PR since the DB isn't implemented, but I'm not 100% sure
> 
> These have nothing to do with the DB though.
> 
> Maybe this will be more clear:
> 
> * Circle 1: profile picture and username are misaligned. If anything, the pfp should be centered and not touching the divider.
> * Circle 2: the username is breaking because the container is small, but there is ample amount of space for the username to avoid wrapping.
> * Circle 3: this is the only one involving the DB, where the message in the circle doesn't appear in the chat window.
> * Circle 4: the height of the messaging component doesn't extend to the whole available page.

The text wrapping is most likely because the "time since last message" isn't displayed

**Comment by AlyaNaseer** [comment]
The only thing I noticed is now the text box for the text isn't always at the bottom, as the screen gets bigger. You might need to add the "height:100%" I suggested elsewhere too (like in the message window component)

image

Also, I'd add a placeholder for the last message and for the time, just so we know what it'll look like. I'd also append a "@" to the username here: 
image

**Comment by itsShaVawn** [comment]
> 

the what?

**Comment by itsShaVawn** [comment]
> > > > In addition to what @AlyaNaseer said: image
> > > > I would have also liked to see the "New message" button being used in your video.
> > > 
> > > 
> > > I think some of the changes you ask might be difficult in this PR since the DB isn't implemented, but I'm not 100% sure
> > 
> > 
> > These have nothing to do with the DB though.
> > Maybe this will be more clear:
> > 
> > * Circle 1: profile picture and username are misaligned. If anything, the pfp should be centered and not touching the divider.
> > * Circle 2: the username is breaking because the container is small, but there is ample amount of space for the username to avoid wrapping.
> > * Circle 3: this is the only one involving the DB, where the message in the circle doesn't appear in the chat window.
> > * Circle 4: the height of the messaging component doesn't extend to the whole available page.
> 
> The text wrapping is most likely because the "time since last message" isn't displayed

the what?

**Comment by AlyaNaseer** [comment]
> > > > > In addition to what @AlyaNaseer said: image
> > > > > I would have also liked to see the "New message" button being used in your video.
> > > > 
> > > > 
> > > > I think some of the changes you ask might be difficult in this PR since the DB isn't implemented, but I'm not 100% sure
> > > 
> > > 
> > > These have nothing to do with the DB though.
> > > Maybe this will be more clear:
> > > 
> > > * Circle 1: profile picture and username are misaligned. If anything, the pfp should be centered and not touching the divider.
> > > * Circle 2: the username is breaking because the container is small, but there is ample amount of space for the username to avoid wrapping.
> > > * Circle 3: this is the only one involving the DB, where the message in the circle doesn't appear in the chat window.
> > > * Circle 4: the height of the messaging component doesn't extend to the whole available page.
> > 
> > 
> > The text wrapping is most likely because the "time since last message" isn't displayed
> 
> the what?

Your Circle 2 text wrapping is most likely due to the fact that the "5m" (time since last message sent) is not visible in this PR
image

**Comment by tungleu** [comment]
> 

image
Addressed :)

**Review by AlyaNaseer — APPROVED** [review]

**Review by itsShaVawn — APPROVED** [review]
👍🏼

**Pull request merged**